File: src/main/java/aoc2019/day11/Panel.java
package aoc2019.day11;

import java.util.HashMap;
import java.util.Map;

public class Panel {

    private int currentPanel;
    private Direction direction;

    public Panel(int currentPanel, Direction direction) {
        this.currentPanel = currentPanel;
        this.direction = direction;
    }

    public int getCurrentPanel() {
        return currentPanel;
    }

    public void paintPanelBlack() {
        this.currentPanel = 0;
    }

    public void paintPanelWhite() {
        this.currentPanel = 1;
    }

    public Direction getDirection() {
        return direction;
    }

    public void setDirection(Direction direction) {
        this.direction = direction;
    }

    enum Direction {
        UP('^'), DOWN('v'), RIGHT('>'), LEFT('<');
        private char dir;
        private static Map map = new HashMap<>();

        Direction(char dir) {
            this.dir = dir;
        }

        public char getDirChar() {
            return this.dir;
        }

        static {
            for (Direction d : Direction.values()) {
                map.put(d.dir, d);
            }
        }

        public static Direction valueOf(char direction) {
            return (Direction) map.get(direction);
        }
    }

    @Override
    public String toString() {
        return "{" +
                "current=" + currentPanel +
                ", direction=" + direction +
                '}';
    }
}

File: 2015-2023/src/main/kotlin/aoc2019/day11/Panel.kt
package aoc2019.day11

import java.util.HashMap

class Panel(var currentPanel: Int, var direction: Direction) {
    fun paintPanelBlack() {
        currentPanel = 0
    }

    fun paintPanelWhite() {
        currentPanel = 1
    }

    enum class Direction(val dirChar: Char) {
        UP('^'), DOWN('v'), RIGHT('>'), LEFT('<');

        companion object {
            private val map: MutableMap<Char, Direction> = HashMap<Char, Direction>()
            fun valueOf(direction: Char): Direction? {
                return map[direction] as Direction?
            }

            init {
                for (d in values()) {
                    map[d.dirChar] = d
                }
            }
        }
    }

    override fun toString(): String {
        return "{" +
                "current=" + currentPanel +
                ", direction=" + direction +
                '}'
    }
}